File: DynamicMapping/dynamic_mapping/geometry_maps/primitives/circle_builder.py
import numpy as np

class CircleBuilder:

    def __init__(self, points : dict | None = None):
        """
        Instantiates a class to build circles from points
        
            :param points: The current points we want to add in the circle
        """

        # Regression parameters of the circle [Rx, Ry]:
        self.Rxy = np.zeros(2)

        # Points representing this circle:
        if points is None:
            self.points = {}
            self.points['positions'] = []
            self.points['covariances'] = []
        
        else:
            self.points = points

            for position in points['positions']:
                self.Rxy += position

        self.center = self.Rxy / len(self.points['positions'])

    def add_point(self, position : np.ndarray, covariance : np.ndarray):    
        """ Adds a point with the given position (x, y) and covariance to the circle """

        self.points['positions'].append(position)
        self.points['covariances'].append(covariance)
        
        self.Rxy += position

        self.center = self.Rxy / len(self.points['positions'])

    def build(self):
        """ Returns a circle instance, built from the points added to this circle builder """

        # Compute the radius of the circle:
        R = self.compute_circle_radius()

        # Build the circle:
        circle = (self.center, R)

        return circle

    def compute_circle_radius(self):
        """ Computes the radius of the circle from the position of the points """

        n = len(self.points['positions'])

        if (n == 1):
            return 0

        sigma = np.zeros(2)
        for position in self.points['positions']:
            sigma += (position - self.center) ** 2
        
        # The division by 'n-1' instead of 'n' is known as Bessel's correction,
        # which corrects the bias in the estimation of the points variance:
        return np.sqrt(np.sum(sigma / (n - 1)))

    def distance_from(self, point):
        """ Returns the distance between the circle center and the given point """
        
        return np.sqrt(np.sum((point[1] - self.center) ** 2))

    def points_count(self):
        """ Returns the number of points currently inside the circle """

        return len(self.points['positions'])

File: IDP - Dynamic Mapping/Assets/Scripts/Mapping/GeometryClustering/Primitives/Builders/CircleBuilder.cs
using System.Collections.Generic;
using UnityEngine;

/// <summary>
/// Class used to build circles from observed points
/// </summary>
public class CircleBuilder {
    
    // Points representing this circle:
    private List<Point> points;

    // Regression parameters of the circle (Rx, Ry):
    private Vector2 Rxy = new Vector2();

    // Current center of the circle:
    private Vector2 center;

    /// <summary>
    /// Instantiates a circle builder, to build circles from observed points
    /// </summary>
    public CircleBuilder() {
        points = new List<Point>();
    }

    /// <summary>
    /// Instantiates a circle builder, to build circles from observed points
    /// </summary>
    /// <param name="points">The points to add to the current circle</param>
    public CircleBuilder(List<Point> points) {
        this.points = points;

        foreach (Point point in points)
            Rxy += point.position;

        center = Rxy / points.Count;
    }

    /// <summary>
    /// Adds a point to the circle to build
    /// </summary>
    /// <param name="point">The point to add</param>
    public void AddPoint(Point point) {
        points.Add(point);

        Rxy += point.position;

        center = Rxy / points.Count;
    }

    /// <summary>
    /// Builds a circle from the points added to this circle builder
    /// </summary>
    /// <returns>The circle that was built</returns>
    public Circle Build() {
        // Compute the radius of the circle:
        float R = ComputeCircleRadius();

        // Build the circle:
        Circle circle = new Circle(center, R);

        return circle;
    }

    /// <summary>
    /// Computes the radius of the circle from the position of the points in the circle builder
    /// </summary>
    /// <returns>The radius of the circle</returns>
    private float ComputeCircleRadius() {
        int n = points.Count;

        if (n == 1)
            return 0;

        float sigma_x2 = 0, sigma_y2 = 0;
        foreach (Point p in points) {
            sigma_x2 += (p.position.x - center.x) * (p.position.x - center.x);
            sigma_y2 += (p.position.y - center.y) * (p.position.y - center.y);
        }

        // The division by 'n-1' instead of 'n' is known as Bessel's correction,
        // which corrects the bias in the estimation of the points variance:
        sigma_x2 /= n - 1; sigma_y2 /= n - 1;

        return Mathf.Sqrt(sigma_x2 + sigma_y2);
    }

    /// <summary>
    /// Returns teh distance between the circle center and the given point
    /// </summary>
    public float DistanceFrom(Point point) {
        return (point.position - center).magnitude;
    }

    /// <summary>
    /// Returns the number of points currently in the circle
    /// </summary>
    public int PointsCount() {
        return points.Count;
    }
}
